org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServers.java

```diff
@@ -268,6 +282,72 @@ public long getStartVersion() {
 			return this.startVersion;
 		}
 
+		public SingleLanguageServerDocumentExecutor toExecutor(final @NonNull LanguageServerWrapper serverWrapper) {
+			return new SingleLanguageServerDocumentExecutor(this.document, serverWrapper);
+		}
+
+		@Override
+		public LanguageServerDocumentExecutor clone() {
+			return new LanguageServerDocumentExecutor(this.document).withFilter(getFilter());
+		}
+
+		@Override
+		public void reset() {
+			super.reset();
+			this.startVersion = 0L;
+		}
+	}
+
+	/**
+	 * Executor for a single server - for follow-up requests that should go to the server that returned an earlier response.
+	 * The multi-LS dispatch methods will work, and will take into account the filter, so will return exactly 1 or 0 response.
+	 * The additional <code>execute()</code> method will bypass the filter and just run the request directly on the single wrapped
+	 * language server.
+	 *
+	 */
+	public static class SingleLanguageServerDocumentExecutor extends LanguageServerDocumentExecutor {
+
+		private final @NonNull LanguageServerWrapper serverWrapper;
+
+		SingleLanguageServerDocumentExecutor(final @NonNull IDocument document, final @NonNull LanguageServerWrapper serverWrapper) {
+			super(document);
+			this.serverWrapper = serverWrapper;
+		}
+
+		@Override
+		@SuppressWarnings("null")
+		protected @NonNull List<@NonNull CompletableFuture<@Nullable LanguageServerWrapper>> getServers() {
+			// Take into account filter
+ 			if (getFilter().test(serverWrapper.getServerCapabilities())) {
+ 				return Collections.singletonList(CompletableFuture.completedFuture(serverWrapper));
+ 			}
+			return Collections.emptyList();
+		}
+
+		/**
```

Comment:
I think we should remove this three methods for the time being and add then later if needed, it is really weird that they promote different patterns, like an executor which does not respect the filter.

I would propose that we go without `execute`, `supports` and `getServer`, change a consumers, and the create a new PR showing how those consumers would be indeed simplified.

Could we do that? Adding the methods is indeed simple if really needed.

Reply:
The methods were all driven by necessity - the design choices were prompted by the issues I encountered porting code. See for example `CodeActionCompletionProposal.apply()`.

I like the idea that if you want to follow up your initial request with a further action - but only if the server supports the more refined capability - then you can call `clone().withFilter(...).computeFirst()` and get a result only if the server supports it.

In the case of trying to port the apply method above, there are two criteria for proceeding, only one of which is a server predicate, so just using the filter method doesn't work.